hivemq-edge/src/frontend/src/extensions/datahub/flow/transition/TransitionPanel.tsx

```diff
@@ -1,29 +1,108 @@
+import {
+  BehaviorPolicyData,
+  DataHubNodeType,
+  FiniteStateMachineSchema,
+  PanelProps,
+  StateType,
+  TransitionData,
+  TransitionType,
+} from '@datahub/types.ts'
 import useDataHubDraftStore from '@datahub/hooks/useDataHubDraftStore.ts'
 import { ReactFlowSchemaForm } from '@datahub/components/forms/ReactFlowSchemaForm.tsx'
 import { MOCK_TRANSITION_SCHEMA } from '@datahub/flow/transition/TransitionData.ts'
+import { datahubRJSFWidgets } from '@datahub/flow/datahubRJSFWidgets.tsx'
+import { MOCK_BEHAVIOR_POLICY_SCHEMA } from '@datahub/flow/behavior_policy/BehaviorPolicySchema.ts'
+import { FiniteStateMachineFlow } from '@datahub/components/fsm/FiniteStateMachineFlow.tsx'
 
 export const TransitionPanel: FC<PanelProps> = ({ selectedNode, onFormSubmit }) => {
-  const { nodes } = useDataHubDraftStore()
+  const { nodes, edges } = useDataHubDraftStore()
+
+  // TODO[NVL] Error messages?
+  const parentPolicy = useMemo(() => {
+    const adapterNode = nodes.find((e) => e.id === selectedNode) as Node<TransitionData> | undefined
+    if (!adapterNode) return undefined
+    const incomers = getIncomers(adapterNode, nodes, edges)
+    if (!incomers) return undefined
+    if (incomers.length !== 1) return undefined
+    const [behavior] = incomers
+    if (behavior.type !== DataHubNodeType.BEHAVIOR_POLICY) return undefined
+
+    return behavior as Node<BehaviorPolicyData>
+  }, [edges, nodes, selectedNode])
 
   const data = useMemo(() => {
     const adapterNode = nodes.find((e) => e.id === selectedNode) as Node<TransitionData> | undefined
-    return adapterNode ? adapterNode.data : null
-  }, [selectedNode, nodes])
+    if (!adapterNode) return null
+
+    const { event, from, to } = adapterNode.data
+    const tempData: TransitionData = {
+      ...adapterNode.data,
+      model: parentPolicy ? parentPolicy.data.model : undefined,
+      // @ts-ignore
+      event: `${event || ''}-${from || ''}-${to || ''}`,
+    }
+    return tempData
+  }, [nodes, parentPolicy, selectedNode])
+
+  const options = useMemo(() => {
+    if (!parentPolicy) return null
+    const { model } = parentPolicy.data
+
+    const definition = MOCK_BEHAVIOR_POLICY_SCHEMA.schema.definitions?.[model]
+    if (!definition) return null
+
+    const { metadata } = definition as FiniteStateMachineSchema
+    const { states, transitions } = metadata
+    if (!states || !transitions) return null
+
+    return metadata
+  }, [parentPolicy])
+
+  const onSafeFormSubmit = useCallback(
+    (data: IChangeEvent<TransitionData>) => {
+      const initData = data
+      const { formData } = initData
+      if (formData) {
+        const { event } = formData
+        if (event) {
+          const [e, from, to] = event.split('-')
+          initData.formData = {
+            ...initData.formData,
+            event: e as unknown as TransitionType,
+            from: from as unknown as StateType,
+            to: to as unknown as StateType,
```

Comment:
suggestion: You may can reduce the amount of manual assignments. WDYT about the following

 ```suggestion
           const [event, from, to] = event.split('-') as unknown as [TransitionType, StateType, StateType]
           initData.formData = {
             ...initData.formData,
             event,
             from,
             to,
```

Reply:
Do you find that more readable? It's leaner but ambivalent as to whether it reads better. I'll try it, see if it works for me

Reply:
It works better, thanks to implicit typing and the IDE. And `unknown` is not even needed there